doc/reference/trd-userspace-readable-allow-syscalls.md

```diff
@@ -0,0 +1,175 @@
+Userspace Readable Allow System Call
+========================================
+
+**TRD:** XXX <br/>
+**Working Group:** Kernel<br/>
+**Type:** Documentary<br/>
+**Status:** Draft <br/>
+**Author:** Alistair Francis <br/>
+**Draft-Created:** June 17, 2021<br/>
+**Draft-Modified:** June 17, 2021<br/>
+**Draft-Version:** 1<br/>
+**Draft-Discuss:** tock-dev@googlegroups.com</br>
+
+Abstract
+-------------------------------
+
+This document describes the userspace readable allow system call application binary
+interface (ABI) between user space processes and the Tock kernel for 32-bit
+ARM Cortex-M and RISC-V RV32I platforms.
+
+This is an extension on the allow calls specified in
+[TRD 104](trd104-syscalls.md).
+
+1 Introduction
+==============
+
+In normal use, userspace does not access a buffer that
+has been userspace readable with the kernel with a Read-Write Allow call. This
+reading restriction is because the contents of the buffer may be in an
+intermediate state and so not consistent with expected data
+models. Ensuring every system call driver maintains consistency in the
+presence of arbitrary userspace reads is too great a programming
+burden for an unintended use case.
+
+However, there can be cases when it is necessary for userspace to be
+able to read a buffer without first revoking it from the kernel with a
+Read-Write Allow. These cases are situations when the cost of a
+Read-Write Allow system call is an unacceptable overhead for
+accessing the data.
+
+<<<<<<< HEAD:doc/reference/trd-shared-allow-syscalls.md
+Instead, capsules that support the shared allow call can communicate with
+=======
+This allows capsules that support the userspace readable allow call to communicate with
+>>>>>>> 16c4dfaba (kernel: Rename shared_allow to userspace_readable_allow):doc/reference/trd-userspace-readable-allow-syscalls.md
+applications without buffers needing to be re-allowed. For example a capsule
+might want to report statistics to a userspace app. It could do this by letting
+the app perform a userspace readable allow call to allocate a buffer. Then the capsule can
+write statistics to the buffer and at any time the app can read the statistics
+from the buffer.
+
+2 System Call API
+=================================
+
+2.1 Userspace Readable Allow (Class ID: 7)
+---------------------------------
+The userspace readable allow syscall follows the same expectations and
+requirements as described for the Read-Write syscall in
+[TRD104 Section 4.4](trd104-syscalls.md#44-read-write-allow-class-id-3), with
+the exception that apps are explicitly allowed to read buffers that have
+trd-userspace-readable-allow-syscalls.md
+been passed to the kernel.
+
```

Comment:
Can you add the text here the expectations of userspace reading/writing? The rules are very specific.

I would suggest

"Finally, because a process conceptually relinquishes write access to a buffer when it makes a Read-Write Allow call with it, a userspace API MUST NOT assume or rely on a process writing an allowed buffer. If userspace needs to write to a buffer held by the kernel, it MUST first regain access to it by calling the corresponding Userspace Readable Allow. A userspace API MAY allow a process to read an allowed buffer, but if it does, it must document a consistency mechanism as described below."

In the other allows, this consistency text is later, so it might make sense to just push it later (and remove the last sentence).